### PR TITLE
SageMaker custom types for ParameterRangeOneOf and HyperparameterConfig. TunableParams extraction

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
- [x] Creates custom types for ParameterRangeOneOf and HyperparameterJobConfig.
- [x] Use generics to serialize/deserialize protos to make UX visualization easier
- [x] Move tunable hyperparameters out of HyperparameterJobConfig to make it easier to bind (and consistent with Custom Container Training Job)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

Example:
```
simple_xgboost_hpo_job_task = hpo_job_task.SdkSimpleHyperparameterTuningJobTask(
    training_job=builtin_algorithm_training_job_task2,
    max_number_of_training_jobs=10,
    max_parallel_training_jobs=5,
    cache_version="1",
    retries=2,
    cacheable=True,
    tunable_parameters=["num_round", "max_depth", "gamma"],
)

    a = simple_xgboost_hpo_job_task(
        train=train_dataset,
        validation=validation_dataset,
        static_hyperparameters=static_hyperparameters,
        hyperparameter_tuning_job_config=hyperparameter_tuning_job_config,
        num_round=IntegerParameterRange(min_value=2, max_value=8, scaling_type=HyperparameterScalingType.LINEAR),
        max_depth=IntegerParameterRange(min_value=5, max_value=7, scaling_type=HyperparameterScalingType.LINEAR),
        gamma=ContinuousParameterRange(min_value=0.0, max_value=0.3, scaling_type=HyperparameterScalingType.LINEAR),
    )
```

## Tracking Issue
https://github.com/lyft/flyte/issues/455

### CODE REVIEW

File: tests/flytekit/unit/common_tests/types/test_proto.py [hunk 1]
```diff
@@ -26,7 +27,16 @@ def test_proto_to_literal_type():
 def test_proto():
     proto_type = _proto.create_protobuf(_errors_pb2.ContainerError)
     assert proto_type.short_class_string() == "Types.Proto(flyteidl.core.errors_pb2.ContainerError)"
+    run_test_proto_type(proto_type)
```
Comment (wild-endeavor) [inline tests/flytekit/unit/common_tests/types/test_proto.py:30]: Could we get a unit test for `get_sdk_type_from_literal_type`?

**Review by wild-endeavor — APPROVED** [review]
I'd like more time to look into this but I'm headed off tomorrow so kinda approving with some comments that you can ignore if you don't want to address them.

File: flytekit/sdk/types.py [hunk 2]
```diff
@@ -395,6 +395,47 @@ def assert_and_create(wf_params, a, b):
             )
     """
 
+    GenericProto = staticmethod(_proto.create_generic)
+    """
+    Use this to specify a custom protobuf type that will get serialized as a proto struct. This enables UX visualization
```
[the remark this answers is not in the document]
Comment (EngHabu) [reply flytekit/sdk/types.py]: how about now?

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 3]
```diff
@@ -185,3 +193,73 @@ def from_flyte_idl(cls, pb2_object: _idl_parameter_ranges.ParameterRanges):
...
+        """
+        self._integer_parameter_range = param if isinstance(param, IntegerParameterRange) else None
+        self._continuous_parameter_range = param if isinstance(param, ContinuousParameterRange) else None
+        self._categorical_parameter_range = param if isinstance(param, CategoricalParameterRange) else None
+
+    @property
+    def integer_parameter_range(self) -> Optional[IntegerParameterRange]:
+        """
+        Retrieves the integer parameter range if one is set. None otherwise.
+        :rtype: Optional[IntegerParameterRange]
```
[the remark this answers is not in the document]
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py:227]: What's "the other thing" here?

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 4]
```diff
@@ -177,8 +178,15 @@ def to_flyte_idl(self) -> _idl_parameter_ranges.ParameterRanges:
                 converted[k] = _idl_parameter_ranges.ParameterRangeOneOf(integer_parameter_range=v.to_flyte_idl())
             elif isinstance(v, ContinuousParameterRange):
                 converted[k] = _idl_parameter_ranges.ParameterRangeOneOf(continuous_parameter_range=v.to_flyte_idl())
-            else:
+            elif isinstance(v, CategoricalParameterRange):
```
Comment (bnsblue) [inline flytekit/models/sagemaker/parameter_ranges.py:181]: Can you pull Master? This check actually doesn't work as expected. I have merged my fix.

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py:227]: @wild-endeavor ^

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 5]
```diff
@@ -199,3 +207,73 @@ def from_flyte_idl(cls, pb2_object: _idl_parameter_ranges.ParameterRanges):
...
+            continuous_parameter_range=self.continuous_parameter_range.to_flyte_idl()
+            if self.continuous_parameter_range
+            else None,
+            categorical_parameter_range=self.categorical_parameter_range.to_flyte_idl()
+            if self.categorical_parameter_range
+            else None,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb_object: _idl_parameter_ranges.ParameterRangeOneOf):
```
Comment (bnsblue) [inline flytekit/models/sagemaker/parameter_ranges.py:270]: Oh I see. I was missing this intermediate layer. That's why my `ParameterRanges.from_flyte_idl()` didn't work as expected.

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 6]
```diff
@@ -199,3 +207,73 @@ def from_flyte_idl(cls, pb2_object: _idl_parameter_ranges.ParameterRanges):
                 converted[k] = CategoricalParameterRange.from_flyte_idl(v.categorical_parameter_range)
 
         return cls(parameter_range_map=converted,)
+
+
+class ParameterRangeOneOf(_common.FlyteIdlEntity):
```
Comment (bnsblue) [inline flytekit/models/sagemaker/parameter_ranges.py:212]: Could you add a unit test for this? I added a set of unittests for SageMaker models in my latest PR

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 4 above)
Comment (EngHabu) [reply flytekit/models/sagemaker/parameter_ranges.py:181]: What are you referring to exactly? [this](https://github.com/lyft/flytekit/blob/master/flytekit/models/sagemaker/parameter_ranges.py#L169-L183) is the copy on master

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply flytekit/models/sagemaker/parameter_ranges.py:227]: rtype "comment" vs py3 typing hint (the `-> Optional[IntegerParameterRange]` a few lines above)

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 6 above)
Comment (EngHabu) [reply flytekit/models/sagemaker/parameter_ranges.py:212]: Done 👍

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py:227]: ohh got it

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 4 above)
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py:181]: My bad. I thought this was `from_flyte_idl()`

Comment (EngHabu) [reply flytekit/models/sagemaker/parameter_ranges.py:181]: So, all good?

**Comment by EngHabu** [comment]
Ping @wild-endeavor @bnsblue

File: flytekit/sdk/types.py [hunk 7]
```diff
@@ -395,6 +419,52 @@ def assert_and_create(wf_params, a, b):
...
+        As input:
+            1) If set, a Python protobuf object of the type specified in the definition.
+            2) If not set, a None value.
+
+        As output:
+            1) A Python protobuf object matching the type specified by the users.
+            2) Set None to null the output.
+
+        From command-line:
+            A base-64 encoded string of the serialized protobuf.
```
Comment (bnsblue) [inline flytekit/sdk/types.py:448]: nit: I am not sure if I understand what this is trying to convey

File: tests/flytekit/common/workflows/sagemaker.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            training_job_early_stopping_type=TrainingJobEarlyStoppingType.AUTO,
+        ),
+    )
+
+    a = simple_xgboost_hpo_job_task(
+        train=train_dataset,
+        validation=validation_dataset,
+        static_hyperparameters=static_hyperparameters,
+        hyperparameter_tuning_job_config=hyperparameter_tuning_job_config,
+        num_round=ParameterRangeOneOf(
```
Comment (bnsblue) [inline tests/flytekit/common/workflows/sagemaker.py:100]: Why is it `ParameterRangeOneOf` instead of `ParameterRange` as defined here? https://github.com/lyft/flytekit/pull/189/files#diff-0a303f3854a09b767e67d5f6eb857cbc49a6714f8df987f65646b98c5ef4d194R8

Comment (bnsblue) [inline tests/flytekit/common/workflows/sagemaker.py:100]: Also, is it possible to make this even simpler by getting rid of the `ParameterRangeOneOf()` part? For example,  Is it possible to make it as simple as `num_round=IntegerParameterRange(...)` ?

Comment (bnsblue) [reply tests/flytekit/common/workflows/sagemaker.py:100]: What's the purpose of `flytekit.common.tasks.sagemaker.types.ParameterRange`?

Comment (bnsblue) [inline tests/flytekit/common/workflows/sagemaker.py:100]: Sorry I don't think I understand the code so these might just be stupid questions. But when should the user use ParameterRangeOneOf and when should one use ParameterRange?

**Comment by bnsblue** [comment]
I left some questions and comments. I don't think I understand the code fully so those questions might be stupid.

File: flytekit/sdk/types.py (hunk 7 above)
Comment (EngHabu) [reply flytekit/sdk/types.py:448]: If you specify this type as an input parameter or a Workflow and you want to launch the workflow from CMD (like through `flyte-cli execute-launch-plan`) this tells you how to pass the value

File: tests/flytekit/common/workflows/sagemaker.py (hunk 8 above)
Comment (EngHabu) [reply tests/flytekit/common/workflows/sagemaker.py:100]: That (in common.tasks.sagemaker.types) is just the "type" definition.. this is the value...
just like you use `Types.String` to define the type of an input but you then use `"hello world"` as the value...

Comment (EngHabu) [reply tests/flytekit/common/workflows/sagemaker.py:100]: I would love to... any ideas? @wild-endeavor might have ideas too..

Comment (EngHabu) [reply tests/flytekit/common/workflows/sagemaker.py:100]: ParameterRange is the type... you use it when you declare the interface of a task/WF... etc...
ParameterRangeOneOf is the value you can set for that type.... 

We can choose to name them the same but keep them in two different packages... (the flytekit way as far as I understand)...

@wild-endeavor  might help shed some light on how to properly do this...

File: flytekit/sdk/types.py (hunk 7 above)
Comment (bnsblue) [reply flytekit/sdk/types.py:448]: I see

File: tests/flytekit/common/workflows/sagemaker.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        instance_type="ml.m4.xlarge", instance_count=1, volume_size_in_gb=25,
+    ),
+    algorithm_specification=AlgorithmSpecification(
+        input_mode=InputMode.FILE,
+        input_content_type=InputContentType.TEXT_CSV,
+        algorithm_name=AlgorithmName.XGBOOST,
+        algorithm_version="0.72",
+    ),
+)
+
```
Comment (bnsblue) [inline tests/flytekit/common/workflows/sagemaker.py:67]: Do you not need to register the tasks first? I took the content of this file and ran it and encountered this error
```
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = "failed to compile workflow for [resource_type:WORKFLOW project:"flyteexamples" domain:"development" name:"SageMakerHPO" version:"with_custom_type1" ] with err entry not found"
	debug_error_string = "{"created":"@1602714149.782035000","description":"Error received from peer ipv6:[::1]:30081","file":"src/core/lib/surface/call.cc","file_line":1062,"grpc_message":"failed to compile workflow for [resource_type:WORKFLOW project:"flyteexamples" domain:"development" name:"SageMakerHPO" version:"with_custom_type1" ] with err entry not found","grpc_status":13}"
>
```
which was fixed after I register the two tasks first.

File: tests/flytekit/common/workflows/sagemaker.py (hunk 8 above)
Comment (bnsblue) [reply tests/flytekit/common/workflows/sagemaker.py:100]: ah I see

Comment (bnsblue) [reply tests/flytekit/common/workflows/sagemaker.py:100]: pinging @wild-endeavor ^ please take a look

**Comment by bnsblue** [comment]
I don't have further comments besides I really like the user to be able to write 
`num_round=IntegerParameterRange(...)` instead of `num_round=ParameterRangeOneOf(IntegerParameterRange(...))`
https://github.com/lyft/flytekit/pull/189#discussion_r504410081
Unfortunately I am not able to come up with a way to solve it, so I am going to approve the PR.

**Review by bnsblue — DISMISSED** [review]

**Comment by EngHabu** [comment]
Let me try something

**Comment by bnsblue** [comment]
Sure. I was hoping that creating a ParameterRangeBase and letting ParameterRangeOneOf and IntegerParameterRange etc all inherit from it can solve the problem, but I couldn't get it to work.

**Comment by bnsblue** [comment]
Are you done with what you wanted to try? Is that part ready for review?

File: flytekit/common/tasks/sagemaker/types.py [hunk 10]
```diff
@@ -1,8 +1,6 @@
...
-from flyteidl.plugins.sagemaker import parameter_ranges_pb2 as _pb2_parameter_ranges
-
+from flytekit.models.sagemaker import parameter_ranges as _parameter_range_models, hpo_job as _hpo_models
 from flytekit.sdk import types as _sdk_types
 
-HyperparameterTuningJobConfig = _sdk_types.Types.GenericProto(_pb2_hpo_job.HyperparameterTuningJobConfig)
+HyperparameterTuningJobConfig = _sdk_types.Types.GenericProto(_hpo_models.HyperparameterTuningJobConfig)
 
-ParameterRange = _sdk_types.Types.GenericProto(_pb2_parameter_ranges.ParameterRangeOneOf)
+ParameterRange = _sdk_types.Types.GenericProto(_parameter_range_models.ParameterRangeOneOf)
```
Comment (bnsblue) [inline flytekit/common/tasks/sagemaker/types.py:7]: If we are wrapping a `model` instead, should we still call it `Generic"Proto"`?

**Comment by bnsblue** [comment]
just a nit. LGTM

**Review by bnsblue — APPROVED** [review]

File: flytekit/common/tasks/sagemaker/types.py (hunk 10 above)
Comment (EngHabu) [reply flytekit/common/tasks/sagemaker/types.py:7]: models should be thought of as "glorified" protos (or pythonic-protos)... so same assumptions about them should apply...

Comment (bnsblue) [reply flytekit/common/tasks/sagemaker/types.py:7]: makes sense
